src/Database/DBStructure.php

```diff
@@ -0,0 +1,1689 @@
+			$admin_mail_list
+		);
+
+		// No valid result?
+		if (!DBM::is_result($adminlist)) {
+			logger(sprintf('Cannot notify administrators about update_id=%d, error_message=%s', $update_id, $error_message), LOGGER_NORMAL);
+
+			// Don't continue
+			return;
+		}
+
+		// every admin could had different language
+		foreach ($adminlist as $admin) {
+			$lang = (($admin['language'])?$admin['language']:'en');
+			push_lang($lang);
+
+			$preamble = deindent(t("
+				The friendica developers released update %s recently,
+				but when I tried to install it, something went terribly wrong.
+				This needs to be fixed soon and I can't do it alone. Please contact a
+				friendica developer if you can not help me on your own. My database might be invalid."));
+			$body = t("The error message is\n[pre]%s[/pre]");
+			$preamble = sprintf($preamble, $update_id);
+			$body = sprintf($body, $error_message);
+
+			notification(array(
+				'type' => SYSTEM_EMAIL,
+				'to_email' => $admin['email'],
+				'preamble' => $preamble,
+				'body' => $body,
+				'language' => $lang)
+			);
+		}
+
+		//try the logger
+		logger("CRITICAL: Database structure update failed: ".$error_message);
+	}
+
+
+	private static function tableStructure($table) {
+		$structures = q("DESCRIBE `%s`", $table);
+
+		$full_columns = q("SHOW FULL COLUMNS FROM `%s`", $table);
+
+		$indexes = q("SHOW INDEX FROM `%s`", $table);
+
+		$table_status = q("SHOW TABLE STATUS WHERE `name` = '%s'", $table);
+
+		if (DBM::is_result($table_status)) {
+			$table_status = $table_status[0];
+		} else {
+			$table_status = array();
+		}
+
+		$fielddata = array();
+		$indexdata = array();
+
+		if (DBM::is_result($indexes)) {
+			foreach ($indexes AS $index) {
+				if ($index['Key_name'] != 'PRIMARY' && $index['Non_unique'] == '0' && !isset($indexdata[$index["Key_name"]])) {
+					$indexdata[$index["Key_name"]] = array('UNIQUE');
+				}
+
+				$column = $index["Column_name"];
+
+				if ($index["Sub_part"] != "") {
+					$column .= "(".$index["Sub_part"].")";
+				}
+
+				$indexdata[$index["Key_name"]][] = $column;
+			}
+		}
+		if (DBM::is_result($structures)) {
+			foreach ($structures AS $field) {
+				$fielddata[$field["Field"]]["type"] = $field["Type"];
+				if ($field["Null"] == "NO") {
+					$fielddata[$field["Field"]]["not null"] = true;
+				}
+
+				if (isset($field["Default"])) {
+					$fielddata[$field["Field"]]["default"] = $field["Default"];
+				}
+
+				if ($field["Extra"] != "") {
+					$fielddata[$field["Field"]]["extra"] = $field["Extra"];
+				}
+
+				if ($field["Key"] == "PRI") {
+					$fielddata[$field["Field"]]["primary"] = true;
+				}
+			}
+		}
+		if (DBM::is_result($full_columns)) {
+			foreach ($full_columns AS $column) {
+				$fielddata[$column["Field"]]["Collation"] = $column["Collation"];
+			}
+		}
+
+		return array("fields" => $fielddata, "indexes" => $indexdata, "table_status" => $table_status);
+	}
+
+	public static function printStructure($database) {
```

Comment:
The parameter is superfluous since we're printing the structure from `DBStructure::structure()` anyway.

Reply:
I must confess, I hadn't had a look at functionality here.

Reply:
Done